### ISSUE TITLE
(redis 3.0.0 and 3.0.0.post1) AttributeError: 'float' object has no attribute 'items'

### ISSUE THREAD

**Issue by dos09** [andymccurdy/redis-py#1068]
This happened after using redis 3.0.0.post1, when I uninstalled it and installed 2.10.6, it's working.

Traceback (most recent call last):
  File "/myenv/lib/python3.5/site-packages/celery/worker/worker.py", line 205, in start
    self.blueprint.start(self)
  File "/myenv/lib/python3.5/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/myenv/lib/python3.5/site-packages/celery/bootsteps.py", line 369, in start
    return self.obj.start()
  File "/myenv/lib/python3.5/site-packages/celery/worker/consumer/consumer.py", line 317, in start
    blueprint.start(self)
  File "/myenv/lib/python3.5/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/myenv/lib/python3.5/site-packages/celery/worker/consumer/consumer.py", line 593, in start
    c.loop(*c.loop_args())
  File "/myenv/lib/python3.5/site-packages/celery/worker/loops.py", line 91, in asynloop
    next(loop)
  File "/myenv/lib/python3.5/site-packages/kombu/asynchronous/hub.py", line 354, in create_loop
    cb(*cbargs)
  File "/myenv/lib/python3.5/site-packages/kombu/transport/redis.py", line 1040, in on_readable
    self.cycle.on_readable(fileno)
  File "/myenv/lib/python3.5/site-packages/kombu/transport/redis.py", line 337, in on_readable
    chan.handlers[type]()
  File "/myenv/lib/python3.5/site-packages/kombu/transport/redis.py", line 724, in _brpop_read
    self.connection._deliver(loads(bytes_to_str(item)), dest)
  File "/myenv/lib/python3.5/site-packages/kombu/transport/virtual/base.py", line 983, in _deliver
    callback(message)
  File "/myenv/lib/python3.5/site-packages/kombu/transport/virtual/base.py", line 632, in _callback
    self.qos.append(message, message.delivery_tag)
  File "/myenv/lib/python3.5/site-packages/kombu/transport/redis.py", line 149, in append
    pipe.zadd(self.unacked_index_key, time(), delivery_tag) \
  File "/myenv/lib/python3.5/site-packages/redis/client.py", line 2263, in zadd
    for pair in iteritems(mapping):
  File "/myenv/lib/python3.5/site-packages/redis/_compat.py", line 123, in iteritems
    return iter(x.items())
AttributeError: 'float' object has no attribute 'items'

**Comment by jpunwin** [comment]
+1

**Comment by andymccurdy** [comment]
Looks like kombu will need to update its code that interfaces with redis-py. Specifically the calls to zadd

**Comment by timthelion** [comment]
Is it really a good idea to make breaking changes like this?

**Comment by gleblanc1783** [comment]
+1 this broke our Airflow instance this morning. Locked in 2.10.6 fixes it for us too.

**Comment by jachicao** [comment]
django-redis might need an update too then.

I'm getting this error when using django.core.cache.get(key)
Invalid input of type: 'CacheKey'. Convert to a byte, string or number first.

**Comment by timthelion** [comment]
@andymccurdy I appreciate the hard work that you put into making a product that my organization relies upon and there is not much we can do about this now. I don't want to beat you up to much about this ;)

So Thank you! :heart: 

That said, it wasn't actually strictly necessary to make the breaking change. It would have been possible to make a new method `zadd2` which would have had the new signature and changed the original `zadd` to call `zadd2` with the correct arguments. Alternatively, it would have been possible to create a method `set_zadd_ops(nx=False, xx=False, ch=False, incr=False)`. Either of these options would have been preferable.

With the chosen solution, it is difficult to support both redis-py 2 and 3. Since python is incapable of using two versions of the same dependency within a given project (admittedly this is a failing of python as much as it is of reddis-py 3). This means that libraries which support reddis-py 2* will necessarily be incompatible with libraries that support reddis 3 and vice versa and it will be impossible to use such libraries together in the same project. That's a pity.

**Comment by andymccurdy** [comment]
I hear you. I take backwards incompatibly very seriously and try to refrain from it when possible.

I agree that adding new methods could have also worked. However, redis-py tried similar things in the past (e.g., Redis vs StrictRedis) and it ended up causing way more problems than it fixed.

I think that other libraries that rely on Redis can actually support both redis-py 2 and 3 with a few lines of code, though I admit it's not the prettiest thing: 

```
import redis
r = redis.Redis(...)

if redis.VERSION[0] < 3:
    r.zadd('my-key', element1=score1)
else:
    r.zadd('my-key', {elemetn1: score1})
```

**Comment by tzickel** [comment]
There is also a problem with libraries which depend on redis-py (you can only have one version installed in a repo), another option is to simply revert redis-py to 2.X and create a new library called redis3-py or redis-py3.

**Comment by jonathanbull** [comment]
Eek, this has also broken our production code. Thanks for providing the workaround @andymccurdy.

**Comment by edmorley** [comment]
> Eek, this has also broken our production code.

Transitive Python dependencies should always be pinned, particularly in production (the same principal as always using a lockfile in the JS ecosystem). This can be enforced using either something like pipenv or [poetry](https://github.com/sdispater/poetry), or else using requirements files and pip >=8's [hash checking mode](https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode).

**Comment by jonathanbull** [comment]
> > Eek, this has also broken our production code.
> 
> Transitive Python dependencies should always be pinned, particularly in production (the same principal as always using a lockfile in the JS ecosystem). This can be enforced using either something like pipenv or [poetry](https://github.com/sdispater/poetry), or else using requirements files and pip >=8's [hash checking mode](https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode).

You're 100% right. Appreciate the pointer.

**Comment by andymccurdy** [comment]
This has been resolved in celery/kombu#953

**Comment by CloudMarc** [comment]
This could have been done in a way that maintained backward compatibility.  See e.g. what the ruby folks did:  https://www.rubydoc.info/github/ezmobius/redis-rb/Redis:zadd
